WordPress/Sniffs/PHP/DisallowAlternativePHPTagsSniff.php

```diff
@@ -0,0 +1,191 @@
+<?php
+/**
+ * WordPress Coding Standard.
+ *
+ * @category PHP
+ * @package  PHP_CodeSniffer
+ * @link     https://make.wordpress.org/core/handbook/best-practices/coding-standards/
+ */
+
+/**
+ * Verifies that no alternative PHP open tags are used.
+ *
+ * If alternative PHP open tags are found, this sniff can fix both the open and close tags.
+ *
+ * @link     https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/580
+ *
+ * @category PHP
+ * @package  PHP_CodeSniffer
+ * @author   Juliette Reinders Folmer <wpplugins_nospam@adviesenzo.nl>
+ */
+class WordPress_Sniffs_PHP_DisallowAlternativePHPTagsSniff implements PHP_CodeSniffer_Sniff {
+
+	/**
+	 * Returns an array of tokens this test wants to listen for.
+	 *
+	 * @return array
+	 */
+	public function register() {
+		$tokens = array(
+			T_OPEN_TAG,
+			T_OPEN_TAG_WITH_ECHO,
+		);
+
+		$asp_enabled   = (boolean) ini_get( 'asp_tags' );
+		$short_enabled = (boolean) ini_get( 'short_open_tag' );
+
+		if ( false === $asp_enabled || ( true === $asp_enabled && false === $short_enabled ) ) {
+			$tokens[] = T_INLINE_HTML;
+		}
+
+		return $tokens;
+
+	} // end register()
+
+	/**
+	 * Processes this test, when one of its tokens is encountered.
+	 *
+	 * @param PHP_CodeSniffer_File $phpcsFile The file being scanned.
+	 * @param int                  $stackPtr  The position of the current token
+	 *                                        in the stack passed in $tokens.
+	 *
+	 * @return void
+	 */
+	public function process( PHP_CodeSniffer_File $phpcsFile, $stackPtr ) {
+		$tokens        = $phpcsFile->getTokens();
+		$openTag       = $tokens[ $stackPtr ];
+		$asp_enabled   = (boolean) ini_get( 'asp_tags' );
+		$short_enabled = (boolean) ini_get( 'short_open_tag' );
```

Comment:
> Why do we base the behavior of the sniff on the ini settings of the PHP build that is running PHPCS? 

If the PHP build on which PHPCS is run has `short_open_tag` on and `asp_tags` on, it will recognize the tags as `T_OPEN_TAG`/ `T_OPEN_WITH_ECHO`.
If these are turned **off**, the tags will be seen as `T_INLINE_HTML`. **_This also goes for any code between the alternative open tags - it will not be tokenized properly! - it will just be seen as `T_INLINE_HTML`! _**

Same goes, of course, for the PHP build on which the code will be run eventually - is the ini settings are on, these tags work, if they are off, they are ignored and just printed as HTML.

So, what I've done, is, make the sniff recognize these kind of tags *independently* of the system settings. That way the tags are 'caught' *independently* of the system PHPCS is being run on.

If the settings are `on`, you will get an `error` as we know for sure that alternative open tags are used.
If the settings are `off` and changes are that these tags are being used, you will get a `warning` saying this needs manual inspection.

> So wouldn't it be better to check for all tags by default, but allow this to be configured in custom rulesets if folks want to allow some of them?

How the `content` of the token is tested needs to be different for the different tokens to make sure that they are being picked up on properly, so I use the `ini` settings to only run the tests which actually makes sense for the system running PHPCS (efficiency).

Letting this be configurable does not make sense as that would in practice be the same as just turning the test off.

> At the very least the settings should be retrieved only once, for better performance.

Good point. I will set up some properties for this and fill these in the `register()` method (only once).